### PR TITLE
Dart support

### PR DESCRIPTION
**Pull request by agent3bood**
This is my first contribution, feedback is welcome.

Release Notes:

- Added Dart language support ([#5343](https://github.com/zed-industries/zed/issues/5343)).

### CODE REVIEW

**Comment by agent3bood** [comment]
@cla-bot check

**Comment by Aarush-Acharya** [comment]
Hey @agent3bood  thanks a lot , was really waiting for it

File: crates/zed/src/languages/dart/highlights.scm
Comment (domesticmouse) [inline crates/zed/src/languages/dart/highlights.scm:222]: Nit: add trailing new line at end of file?

**Comment by domesticmouse** [comment]
Like the CI, I am having issues building:

```console
$ cargo run --release

    Updating crates.io index
    Updating git submodule `git@github.com:tree-sitter/tree-sitter`
error: failed to get `tree-sitter-dart` as a dependency of package `zed v0.122.0 (/Users/brettmorgan/Documents/GitHub/zed/crates/zed)`
```

**Comment by imujtaba8488** [comment]
> This is my first contribution, feedback is welcome.
> 
> Release Notes:
> 
> * Added Dart language support ([#5343](https://github.com/zed-industries/zed/issues/5343)).

Waiting for this to be merged soon, hopefully!

**Comment by agent3bood** [comment]
Building works on my machine, could it be related to CI not having github in known_hosts? Or ssh key not setup?

**Comment by domesticmouse** [comment]
> Building works on my machine, could it be related to CI not having github in known_hosts? Or ssh key not setup?

Yeah, I don't have SSH keys setup. The more interesting question is, why is the CI failing? Surely they have SSH keys setup for the CI bots?

**Comment by agent3bood** [comment]
Made a dork to fix some issues in the original `tree-sitter-dart`, should work now.

**Comment by imujtaba8488** [comment]
Looks like all checks are passed. Could we get it in?

**Comment by sabowsla** [comment]
hey guys what is missing from this just read this like a rollercoster of emotion

**Comment by maxbrunsfeld** [comment]
@agent3bood Thanks for the PR! The language registry API has changed a bit on main. You now need to register the Tree-sitter grammar in a separate call, and then reference it by name in the language `config.toml`. So I think you may need to adjust your code slightly. Can you rebase or merge main, and test this again?

Also, could you include screenshots of the syntax highlighting and outline view?

Just a heads up - we are very soon going to be moving most of Zed's built-in languages into installable extensions (https://github.com/zed-industries/zed/issues/7096). Dart is a major enough language that I'm ok with merging this in the meantime. But once the extension system lands, we will probably remove this code and put it into a new `dart` repository under the new [zed-extensions](https://github.com/orgs/zed-extensions/repositories) GitHub org. We'll probably add you as a collaborator on that repo when we do it, if you're interested in helping maintain it.

**Comment by agent3bood** [comment]
@maxbrunsfeld  done, looking forward for getting this merged and contributing even more to the project.

Screenshot 2024-02-12 at 13 17 46

File: Cargo.toml
Comment (HenriqueNas) [inline Cargo.toml:233]: just for curiosity:
why did you linked your fork of ["official" Dart tree-sitter](https://tree-sitter.github.io/tree-sitter/) repository instead of it self?

Comment (agent3bood) [reply Cargo.toml:233]: There is no official package for Dart, the website links to "https://github.com/UserNobody14/tree-sitter-dart", which I forked to fix some issues on it.

File: crates/zed/src/languages/dart/highlights.scm
Comment (maxbrunsfeld) [inline crates/zed/src/languages/dart/highlights.scm:6]: I think this needs to be `function` or `function.method` for us to pick it up.

File: crates/zed/src/languages/dart/highlights.scm
Comment (maxbrunsfeld) [inline crates/zed/src/languages/dart/highlights.scm:85]: Same with these usages of `@method`.

File: crates/zed/src/languages/dart/highlights.scm
Comment (maxbrunsfeld) [inline crates/zed/src/languages/dart/highlights.scm:128]: I don't think we give special highlighting `@parameter`, because in general, we don't have a way of highlighting all of the occurrences of the parameter consistently. Could you remove these parameter ones?

File: crates/zed/src/languages/dart/highlights.scm
Comment (maxbrunsfeld) [inline crates/zed/src/languages/dart/highlights.scm:155]: I don't think we highlight this group either. Maybe just use `keyword` for these? Or if you want to allow themes to target them specifically, you could do `keyword.include`.

File: crates/zed/src/languages/dart/highlights.scm
Comment (maxbrunsfeld) [inline crates/zed/src/languages/dart/highlights.scm:216]: Same thing with `conditional`, `repeat`, and `exception`: I think  just use `keyword`. You could add a suffix like `keyword.conditional`, but we usually don't do that in our queries, so I don't think any themes will target that.

File: crates/zed/src/languages/dart/highlights.scm
Comment (maxbrunsfeld) [inline crates/zed/src/languages/dart/highlights.scm:221]: Can you remove this? We usually don't highlight errors in our highlight queries, because sometimes there are false positives. Also, I don't think themes target this.

**Review by maxbrunsfeld — CHANGES_REQUESTED** [review]
I left some suggestions on the highlight query. Also, could you grab a screenshot of the outline query? Thank you!

**Comment by LiamHockley** [comment]
Huge thanks for this effort!

As soon as this gets merged in, I am poised to give Zed a try. 
Getting a bit tired of the slowness of VSCode.

**Comment by agent3bood** [comment]
@maxbrunsfeld Updated the highlights as suggested. Also check the file outline view bellow.

Screenshot 2024-02-13 at 11 12 08

**Comment by ebsangam** [comment]
LGTM

**Review by maxbrunsfeld — APPROVED** [review]

**Comment by ebabhi** [comment]
LGTM

**Comment by ebabhi** [comment]
Let's gooo 🚀🚀🚀

**Comment by imujtaba8488** [comment]
> Let's gooo 🚀🚀🚀

repeat > Lets' go 🚀🧨

**Review by HenriqueNas — APPROVED** [review]

**Comment by imujtaba8488** [comment]
> When ship? 🚢🚢🚢

with extension in pre (love it) - Lets' go for Flutter! 🚀🚀🚀

**Comment by sabowsla** [comment]
we all agree that this will change the course of the human timeline right ?

**Comment by erf** [comment]
As Dart developers we're all excited about this. Please stop spamming the issue - and in general only comment on Github issues if you contribute to solving it in some meaningful way.

**Comment by d1y** [comment]
@agent3bood Just my opinion: downgrade to a extensions: https://github.com/zed-industries/extensions

**Comment by agent3bood** [comment]
> @agent3bood Just my opinion: downgrade to a extensions: https://github.com/zed-industries/extensions

Looks like that will be the better option, just need a couple of days to check it out.

**Comment by agent3bood** [comment]
Looks like zed extensions does not support adding language server to the editor, we will wait for this to be merged.

> Languages
Extensions may provide languages to extend Zed with support for a particular language.
Currently this is used for things like syntax highlighting and outlining.
Support for defining language servers in extensions will be coming in the future.

**Comment by maxbrunsfeld** [comment]
Yes, we can move this into an extension later, once we have the extension APIs needed for adding new language servers. Thanks for your work on this @agent3bood.

**Comment by maxbrunsfeld** [comment]
This will ship in Zed 0.124.

**Comment by JannieT** [comment]
> This will ship in Zed 0.124.

I'm on Zed 0.132.2 and there is no Dart support. Am I missing something?

**Pull request merged**
